### PR TITLE
Use riak_core_vnode_master:command_return_vnode when doing vnode folds

### PR DESCRIPTION
**Pull request by Vagabond**
This returns the vnode PID, so if we issue a vnode fold request to a
vnode that has completed handoff and is shutting down, we can monitor
that pid and get a DOWN message and unblock the keylist FSM by throwing
an error.

The vnode fold in the 'bloom fold' code has not yet been addressed.

### CODE REVIEW

File: src/riak_repl_fullsync_helper.erl [hunk 1]
```diff
@@ -151,22 +151,49 @@ handle_call({make_keylist, Partition, Filename}, From, State) ->
...
+                            {Partition, OwnerNode},
+                            Req,
+                            {raw, FoldRef, self()},
+                            riak_kv_vnode_master) of
+                        {ok, VNodePid} ->
+                            lager:info("vnode pid is ~p", [VNodePid]),
+                            MonRef = erlang:monitor(process, VNodePid),
+                            receive
+                                {FoldRef, {Self, _} = Reply} ->
+                                    lager:info("vnode fold reply ~p",
```
Comment (bsparrow435) [inline src/riak_repl_fullsync_helper.erl]: maybe debug level to reduce log spam.

File: src/riak_repl_fullsync_helper.erl [hunk 2]
```diff
@@ -151,22 +151,49 @@ handle_call({make_keylist, Partition, Filename}, From, State) ->
...
+                                               [Reply]),
+                                    %% total is 0, sorry
+                                    riak_core_gen_server:cast(Self,
+                                                              {kl_finish, 0});
+                                {FoldRef, {Self, _, Total} = Reply} ->
+                                    lager:info("vnode fold reply ~p",
```
Comment (bsparrow435) [inline src/riak_repl_fullsync_helper.erl]: Same comment about debug level

File: src/riak_repl_fullsync_helper.erl [hunk 3]
```diff
@@ -151,22 +151,49 @@ handle_call({make_keylist, Partition, Filename}, From, State) ->
...
+                            riak_core_util:make_fold_req(fun ?MODULE:keylist_fold/3,
+                                                         {Self, 0})
+                    end,
+                    try riak_core_vnode_master:command_return_vnode(
+                            {Partition, OwnerNode},
+                            Req,
+                            {raw, FoldRef, self()},
+                            riak_kv_vnode_master) of
+                        {ok, VNodePid} ->
+                            lager:info("vnode pid is ~p", [VNodePid]),
```
Comment (bsparrow435) [inline src/riak_repl_fullsync_helper.erl]: debug level

**Comment by bsparrow435** [comment]
Looks good and is passing replication2. :+1: Some log messages should probably be downgraded to debug to reduce log spam during fullsyncs.

**Comment by Vagabond** [comment]
Yeah, I forgot about the logspam, was left over from debugging. I'll fix it this afternoon.

File: src/riak_repl_aae_source.erl [hunk 4]
```diff
@@ -371,13 +371,32 @@ finish_sending_differences(Bloom, #state{index=Partition}) ->
...
-                                                fun ?MODULE:bloom_fold/3,
-                                                {Self,
-                                                 {serialized, ebloom:serialize(Bloom)}}),
-                             gen_fsm:send_event(Self, diff_done),
-                             ok
-                     end,
+                    FoldRef = make_ref(),
+                    try riak_core_vnode_master:command_return_vnode(
+                            {Partition, OwnerNode},
+                            riak_core_util:make_fold_req(
```
Comment (andrewjstone) [inline src/riak_repl_aae_source.erl:377]: When we talked about it last week, @jtuple and @jrwest were opposed to calling `riak_core_util:make_fold_req` from repl and had me add `riak_kv_vnode:fold/4` so I could pass options instead. This makes the opposite change, so I figured I'd mention it to maximize caremad.

Comment (Vagabond) [inline src/riak_repl_aae_source.erl:377]: @jtuple told me to use riak_core_vnode_master:command_return_vnode yesterday, to cope with 'deleted' vnodes, so I didn't see another way to generate the fold request.

Comment (andrewjstone) [inline src/riak_repl_aae_source.erl:377]: Ok. Once this is merged in, I will add the iterator_refresh option to riak_core_util:make_fold_req/4 calls as appropriate and close #521

**Comment by bsparrow435** [comment]
Passed three runs of replication2 :+1:

**Comment by bsparrow435** [comment]
Passed two runs of riak_repl_aae_fullsync :+1:

**Pull request merged**